File: _379_DesignPhoneDirectory.java
// Time Complexity : o(1) all opratios
// Space Complexity : o(maxNumbers)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach: use assigned and un assigned set before assiging the number by current count, check if unassigned is available
class PhoneDirectory {

    int maxNumbers;
    int count;

    Set<Integer> assigned;
    Set<Integer> unAssigned;

    public PhoneDirectory(int maxNumbers) {
        this.maxNumbers = maxNumbers;
        count = 0;
        assigned = new HashSet<>();
        unAssigned = new HashSet<>();
    }

    public int get() {

        int number = -1;

        if(unAssigned.isEmpty() && count == maxNumbers) return number;

        if(unAssigned.isEmpty()){
            number = count;
            count++;
        }
        else{
            for(int x : unAssigned){
                number = x;
                break;
            }
            unAssigned.remove(number);
        }
        assigned.add(number);
        return number;
    }

    public boolean check(int number) {
        return !assigned.contains(number);
    }

    public void release(int number) {
        if(!check(number)){
           assigned.remove(number);
           unAssigned.add(number);
        }
    }
}

File: 379DesignPhoneDirectory.cpp
/* 
    Time Complexity                              :  O(1) for all the three methods get, check and release
    Space Complexity                             :  O(2*N) to store the numbers in q and their corresponding bool expression in the map mp.
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std; 

// https://leetcode.com/problems/design-phone-directory/

class PhoneDirectory {
private:
    queue<int> q;
    unordered_map<int,bool> mp;
public:
    PhoneDirectory(int maxNumbers) {
        for(int i=0;i<maxNumbers;i++) {
            q.push(i);
            mp[i] = true;
        }
    }
    
    int get() {
        if(q.empty()) 
            return -1;
        
        int val = q.front();
        q.pop();
        mp.erase(val);
        return val;
    }
    
    bool check(int number) {
        if(mp.find(number) == mp.end())
            return false;
        return true;
    }
    
    void release(int number) {
        if(check(number)) return;
        q.push(number);
        mp[number] = true;
    }
};

/**
 * Your PhoneDirectory object will be instantiated and called as such:
 * PhoneDirectory* obj = new PhoneDirectory(maxNumbers);
 * int param_1 = obj->get();
 * bool param_2 = obj->check(number);
 * obj->release(number);
 */